File: model/document/ActWriteOff.js
/**
 * Классы и типы для работы с актами списания.
 * @module module:actWriteOff
 */
import ProductInfo from "../dictionary/ProductInfo"
import {ActWriteOffStatus, ActWriteOffType, Version} from '../types'

/**
 * @class module:actWriteOff.ActWriteOff
 * @classdesc Акт списания со склада.
 * @param {string} uuid Уникальный идентификатор акта списания со склада
 * @param {string} docOwner Кто подает документы
 * @param {?string} identity ИД Акта (клиентский)
 * @param {?number} number Номер акта
 * @param {Date} actDate Дата составления акта
 * @param {?module.actWriteOff#ActWriteOffType} type Причина списания
 * @param {?string} note Примечание
 * @param {module:actWriteOff#ActWriteOffStatus} status Состояние
 * @param {?string} rejectComment Комментарий для отказа на акт списания со склада
 * @param {Version} version Версия протокола ЕГАИС
 * @param {?string} replyId Уникальный идентификатор документа (присваивается УТМ); совпадает с идентификатором исходящего документа, который получили в ответе
 */
export default class ActWriteOff {
    constructor(uuid: string,
                docOwner: string,
                identity: (string | null),
                number: (string | null),
                actDate: Date,
                type: (ActWriteOffType | null),
                note: (string | null),
                status: ActWriteOffStatus,
                rejectComment: (string | null),
                version: Version,
                replyId: (string | null)) {
        this.uuid = uuid;
        this.docOwner = docOwner;
        this.identity = identity;
        this.number = number;
        this.actDate = actDate;
        this.type = type;
        this.note = note;
        this.status = status;
        this.rejectComment = rejectComment;
        this.version = version;
        this.replyId = replyId;
    }
}

/**
 * @class module:actWriteOff.ActWriteOffPosition
 * @classdesc Позиция в акте о списании со склада.
 * @param {string} uuid Уникальный идентификатор
 * @param {string} actWriteOffUuid Идентификатор акта списания, содержащего позицию
 * @param {?string} identity Идентификатор позиции внутри накладной
 * @param {number} quantity Количество
 * @param {?string} informF2RegId Регистрационный номер раздела справки 2
 * @param {?string} informF2MarkInfoJson Информация о марках в формате JSON
 * @param {module:productInfo.ProductInfo} productInfo Информация о продукции
 * @param {string[]} markList Перечень считанных кодов марок
 */
export class ActWriteOffPosition {
    constructor(uuid: string,
                actWriteOffUuid: string,
                identity: (string | null),
                quantity: number,
                informF2RegId: (string | null),
                informF2MarkInfoJson: (string | null),
                productInfo: ProductInfo,
                markList: string[]) {
        this.uuid = uuid;
        this.actWriteOffUuid = actWriteOffUuid;
        this.identity = identity;
        this.quantity = quantity;
        this.informF2RegId = informF2RegId;
        this.informF2MarkInfoJson = informF2MarkInfoJson;
        this.productInfo = productInfo;
        this.markList = markList;
    }
}

File: android/src/main/java/com/evotor/egais/mapper/ActWriteOff.kt
package com.evotor.egais.mapper

import android.database.Cursor
import com.evotor.egais.converter.MarkListConverter
import com.evotor.egais.converter.QuantityBigDecimalConverter
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap

internal fun createActWriteOff(cursor: Cursor): WritableMap {
    val result = Arguments.createMap()
    result.putString("uuid", cursor.getString(cursor.getColumnIndexOrThrow("UUID")))
    result.putString("docOwner", cursor.getString(cursor.getColumnIndexOrThrow("OWNER")))
    result.putString("identity", cursor.getString(cursor.getColumnIndexOrThrow("IDENTITY")))
    result.putString("number", cursor.getString(cursor.getColumnIndexOrThrow("NUMBER")))
    result.putString("actDate", cursor.getString(cursor.getColumnIndexOrThrow("ACT_DATE")))
    result.putString("type", cursor.getString(cursor.getColumnIndexOrThrow("TYPE_WRITE_OFF")))
    result.putString("note", cursor.getString(cursor.getColumnIndexOrThrow("NOTE")))
    result.putString("status", cursor.getString(cursor.getColumnIndexOrThrow("STATUS")))
    result.putString("rejectComment", cursor.getString(cursor.getColumnIndexOrThrow("REJECT_COMMENT")))
    result.putString("version", cursor.getString(cursor.getColumnIndexOrThrow("VERSION")))
    result.putString("replyId", cursor.getString(cursor.getColumnIndexOrThrow("REPLY_ID")))
    return result
}

internal fun createActWriteOffPosition(cursor: Cursor): WritableMap {
    val result = Arguments.createMap()
    result.putString("uuid", cursor.getString(cursor.getColumnIndexOrThrow("UUID")))
    result.putString("actWriteOffUuid", cursor.getString(cursor.getColumnIndexOrThrow("ACT_WRITE_OFF_UUID")))
    result.putString("identity", cursor.getString(cursor.getColumnIndexOrThrow("IDENTITY")))
    result.putDouble("quantity", QuantityBigDecimalConverter.toBigDecimal(cursor.getLong(cursor.getColumnIndexOrThrow("QUANTITY"))).toDouble())
    result.putString("informF2RegId", cursor.getString(cursor.getColumnIndexOrThrow("INFORM_F2_REG_ID")))
    result.putString("informF2MarkInfoJson", cursor.getString(cursor.getColumnIndexOrThrow("INFORM_F2_MARK_INFO_JSON")))
    result.putMap("productInfo", createProductInfo(cursor))
    result.putArray("markList", MarkListConverter.toMarkList(cursor.getString(cursor.getColumnIndexOrThrow("MARK_LIST"))))
    return result
}

internal fun createActWriteOffShop(cursor: Cursor): WritableMap {
    val result = Arguments.createMap()
    result.putString("uuid", cursor.getString(cursor.getColumnIndexOrThrow("UUID")))
    result.putString("docOwner", cursor.getString(cursor.getColumnIndexOrThrow("OWNER")))
    result.putString("identity", cursor.getString(cursor.getColumnIndexOrThrow("IDENTITY")))
    result.putString("number", cursor.getString(cursor.getColumnIndexOrThrow("NUMBER")))
    result.putString("actDate", cursor.getString(cursor.getColumnIndexOrThrow("ACT_DATE")))
    result.putString("type", cursor.getString(cursor.getColumnIndexOrThrow("TYPE_WRITE_OFF")))
    result.putString("note", cursor.getString(cursor.getColumnIndexOrThrow("NOTE")))
    result.putString("status", cursor.getString(cursor.getColumnIndexOrThrow("STATUS")))
    result.putString("rejectComment", cursor.getString(cursor.getColumnIndexOrThrow("REJECT_COMMENT")))
    result.putString("replyId", cursor.getString(cursor.getColumnIndexOrThrow("REPLY_ID")))
    return result
}

internal fun createActWriteOffShopPosition(cursor: Cursor): WritableMap {
    val result = Arguments.createMap()
    result.putString("uuid", cursor.getString(cursor.getColumnIndexOrThrow("UUID")))
    result.putString("actWriteOffShopUuid", cursor.getString(cursor.getColumnIndexOrThrow("ACT_WRITE_OFF_SHOP_UUID")))
    result.putString("identity", cursor.getString(cursor.getColumnIndexOrThrow("IDENTITY")))
    result.putDouble("quantity", QuantityBigDecimalConverter.toBigDecimal(cursor.getLong(cursor.getColumnIndexOrThrow("QUANTITY"))).toDouble())
    result.putString("informF2MarkInfoJson", cursor.getString(cursor.getColumnIndexOrThrow("INFORM_F2_MARK_INFO_JSON")))
    result.putMap("productInfo", createProductInfo(cursor))
    result.putArray("markList", MarkListConverter.toMarkList(cursor.getString(cursor.getColumnIndexOrThrow("MARK_LIST"))))
    return result
}